### PR TITLE
Document cockroach-sql binary

### PR DESCRIPTION
**Pull request by ianjevans**
Fixes DOC-3634.

### CODE REVIEW

**Review by knz — COMMENTED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/14671)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @gemma-shay, @ianjevans, @nickvigilante, and @rail)

___
*[`v22.1/cockroach-sql-binary.md` line 74 at r1](https://reviewable.io/reviews/cockroachdb/docs/14671#-N9p9E5s8MmnfHKbSJIv:-N9p9E5s8MmnfHKbSJIw:blvbg9h) ([raw file](https://github.com/cockroachdb/docs/blob/2fd51389f783763d782c8993208f942466942a12/v22.1/cockroach-sql-binary.md#L74)):*
> ```Markdown
> 
> ~~~ shell
> $ cockroach-sql --execute="<sql statement>;<sql statement>" --execute="<sql-statement>" <flags>
> ```

I personally prefer to teach the short argument form (`-e` / `-f`), your call.
___
*[`v22.1/cockroach-sql-binary.md` line 86 at r1](https://reviewable.io/reviews/cockroachdb/docs/14671#-N9p9L1w9zzLDwqVsnt9:-N9p9L1w9zzLDwqVsntA:b-g2mdvp) ([raw file](https://github.com/cockroachdb/docs/blob/2fd51389f783763d782c8993208f942466942a12/v22.1/cockroach-sql-binary.md#L86)):*
> ```Markdown
> 
> ~~~ sql
> > \q
> ```

let's just document this one.
___
*[`v22.1/cockroach-sql-binary.md` line 126 at r1](https://reviewable.io/reviews/cockroachdb/docs/14671#-N9p9TOL6zjm7DIRa-21:-N9p9TOL6zjm7DIRa-22:b3gm7fe) ([raw file](https://github.com/cockroachdb/docs/blob/2fd51389f783763d782c8993208f942466942a12/v22.1/cockroach-sql-binary.md#L126)):*
> ```Markdown
> `--echo-sql` | Reveal the SQL statements sent implicitly by the command-line utility. For a demonstration, see the [example](#reveal-the-sql-statements-sent-implicitly-by-the-command-line-utility) below.<br><br>This can also be enabled within the interactive SQL shell via the `\set echo` [shell command](#commands).
> <a name="sql-flag-execute"></a> `--execute`<br><br>`-e` | Execute SQL statements directly from the command line, without opening a shell. This flag can be set multiple times, and each instance can contain one or more statements separated by semi-colons. If an error occurs in any statement, the command exits with a non-zero status code and further statements are not executed. The results of each statement are printed to the standard output (see `--format` for formatting options).<br><br>For a demonstration of this and other ways to execute SQL from the command line, see the [example](#execute-sql-statements-from-the-command-line) below.
> `--file <filename>`<br><br>`-f <filename>`<br><br>`< <filename>` |  Read SQL statements from `<filename>`.
> ```

The redirection form with `<` is a completely different feature! And is not part of the CLI interface. Please don't include it here.
___
*[`v22.1/cockroach-sql-binary.md` line 136 at r1](https://reviewable.io/reviews/cockroachdb/docs/14671#-N9pADZXB5rOrLbSiCrh:-N9pADZXB5rOrLbSiCri:b-vj2j9i) ([raw file](https://github.com/cockroachdb/docs/blob/2fd51389f783763d782c8993208f942466942a12/v22.1/cockroach-sql-binary.md#L136)):*
> ```Markdown
> ### Client connection
> 
> {% include {{ page.version.version }}/sql/connection-parameters.md %}
> ```

It would be good to re-order this table to promote `--url` first as the preferred approach, then the discrete parameters as an alternative to the URL.
___
*[`v22.1/cockroach-sql-binary.md` line 148 at r1](https://reviewable.io/reviews/cockroachdb/docs/14671#-N9pA_Ix47wsfUStkj8o:-N9pA_Ix47wsfUStkj8p:b9vtujo) ([raw file](https://github.com/cockroachdb/docs/blob/2fd51389f783763d782c8993208f942466942a12/v22.1/cockroach-sql-binary.md#L148)):*
> ```Markdown
> `cockroach-sql` exhibits different behaviors depending on whether or not the session is interactive and/or whether or not the session outputs on a terminal.
> 
> - A session is **interactive** when `cockroach-sql` is invoked without the `--execute` flag and input is not redirected from a file. In such cases:
> ```

"... without `-e` or `-f`, and the input is a terminal"
___
*[`v22.1/cockroach-sql-binary.md` line 152 at r1](https://reviewable.io/reviews/cockroachdb/docs/14671#-N9pAlBr4LAfIHuFL4G4:-N9pAlBr4LAfIHuFL4G5:b-f0ojlx) ([raw file](https://github.com/cockroachdb/docs/blob/2fd51389f783763d782c8993208f942466942a12/v22.1/cockroach-sql-binary.md#L152)):*
> ```Markdown
>     - The [`check_syntax` option](#sql-option-check-syntax) defaults to `true` if supported by the CockroachDB server (this is checked when the shell starts up).
>     - **Ctrl+C** at the prompt will only terminate the shell if no other input was entered on the same line already.
>     - The shell will attempt to set the `safe_updates` [session variable](set-vars.html) to `true` on the server.
> ```

Missing point: 

"""
- The shell continues to read input after the last command entered.
"""
___
*[`v22.1/cockroach-sql-binary.md` line 158 at r1](https://reviewable.io/reviews/cockroachdb/docs/14671#-N9pAt1r3a3sTi_oX-Tf:-N9pAt1r3a3sTi_oX-Tg:b144qcl) ([raw file](https://github.com/cockroachdb/docs/blob/2fd51389f783763d782c8993208f942466942a12/v22.1/cockroach-sql-binary.md#L158)):*
> ```Markdown
> 
> When a session is both interactive and outputs on a terminal, `cockroach-sql` also activates the interactive prompt with a line editor that can be used to modify the current line of input. Also, command history becomes active.
> ```

This section is missing an explanation on non-interactive behavior, which is different.
___
*[`v22.1/cockroach-sql-binary.md` line 159 at r1](https://reviewable.io/reviews/cockroachdb/docs/14671#-N9pB-nd6T4WFsUL2qpj:-N9pB-nd6T4WFsUL2qpk:b-em4c37) ([raw file](https://github.com/cockroachdb/docs/blob/2fd51389f783763d782c8993208f942466942a12/v22.1/cockroach-sql-binary.md#L159)):*
> ```Markdown
> When a session is both interactive and outputs on a terminal, `cockroach-sql` also activates the interactive prompt with a line editor that can be used to modify the current line of input. Also, command history becomes active.
> 
> ## SQL shell
> ```

Somewhere in this page, mention should be made that the output is part of a stable interface, *except* informational output lines that start with `#` (i.e. users should not make programmatic use of it).
___
*[`v22.1/cockroach-sql-binary.md` line 226 at r1](https://reviewable.io/reviews/cockroachdb/docs/14671#-N9pBIcF45znNDrEYlhV:-N9pBIcG1Frkb1hBG29G:byewme2) ([raw file](https://github.com/cockroachdb/docs/blob/2fd51389f783763d782c8993208f942466942a12/v22.1/cockroach-sql-binary.md#L226)):*
> ```Markdown
> - If you are developing automation that uses the CockroachDB SQL shell, it is more reliable to check for `SQLSTATE` values than for error message strings, which are likely to change.
> 
> ## Examples
> ```

I wonder - could we perhaps share this content with the other page? This way they won't drift from each other as we add more features.
___
*[`v22.1/cockroach-sql-binary.md` line 244 at r1](https://reviewable.io/reviews/cockroachdb/docs/14671#-N9pBDE3CLpd8Xk_njLa:-N9pBDE3CLpd8Xk_njLb:b-91uem4) ([raw file](https://github.com/cockroachdb/docs/blob/2fd51389f783763d782c8993208f942466942a12/v22.1/cockroach-sql-binary.md#L244)):*
> ```Markdown
> {% include_cached copy-clipboard.html %}
> ~~~ shell
> # Using the --url flag:
> ```

Maybe worth demonstrating the URL first.

**Review by rail — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/14671)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @gemma-shay, @ianjevans, and @nickvigilante)

___
*[`v22.1/cockroach-sql-binary.md` line 128 at r1](https://reviewable.io/reviews/cockroachdb/docs/14671#-N9qDtYKDrORD9I3yS3b:-N9qDtYKDrORD9I3yS3c:b-ekj34) ([raw file](https://github.com/cockroachdb/docs/blob/2fd51389f783763d782c8993208f942466942a12/v22.1/cockroach-sql-binary.md#L128)):*
> ```Markdown
> `--file <filename>`<br><br>`-f <filename>`<br><br>`< <filename>` |  Read SQL statements from `<filename>`.
> <a name="sql-flag-format"></a> `--format` | How to display table rows printed to the standard output. Possible values: `tsv`, `csv`, `table`, `raw`, `records`, `sql`, `html`.<br><br>**Default:** `table` for sessions that [output on a terminal](#session-and-output-types); `tsv` otherwise<br /><br />This flag corresponds to the `display_format` [client-side option](#client-side-options).
> `--read-only` | **New in v22.1:** Sets the `default_transaction_read_only` [session variable](show-vars.html#supported-variables) to `on` upon connecting.
> ```

A nit. We do not ship `cockroach-sql` with prior versions, so the note is redundant.



_Code quote:_
```Markdown
**New in v22.1:**
```

File: v22.1/cockroach-sql-binary.md
Comment (nickvigilante) [inline v22.1/cockroach-sql-binary.md:28]:
```suggestion
<div class="filter-content" markdown="1" data-scope="linux">

```

Comment (nickvigilante) [reply v22.1/cockroach-sql-binary.md:28]: I figured out the `copy-clipboard.html` problem. You need to add a line break after each `<div>` call. I think it's a JS thing.

File: v22.1/cockroach-sql-binary.md
Comment (nickvigilante) [inline v22.1/cockroach-sql-binary.md:36]:
```suggestion
<div class="filter-content" markdown="1" data-scope="mac">

```

**Review by nickvigilante — CHANGES_REQUESTED** [review]
Can you replicate this to v22.2 as well? Otherwise, LGTM!

File: v22.1/cockroach-sql-binary.md
Comment (ianjevans) [reply v22.1/cockroach-sql-binary.md:32]: Nice. Done.

File: v22.1/cockroach-sql-binary.md
Comment (ianjevans) [reply v22.1/cockroach-sql-binary.md:42]: Done.

**Comment by ianjevans** [comment]
@knz Incorporated all your feedback.

@rail I removed the inline **new in**, and there's now a notice at the top of the topic.

@nickvigilante I'm so glad you figured the `<div>` issue out. It was driving me crazy.

For ease of review I will move all these changes to the v22.2 files after I get approval.

**Review by rail — APPROVED** [review]
LGTM, :shipit:

File: _includes/v22.1/sql/sql-examples.md
Comment (nickvigilante) [inline _includes/v22.1/sql/sql-examples.md:574]: I'll assume that any reference to Jesse can be kept as an homage.

File: v22.1/cockroach-sql-binary.md
Comment (nickvigilante) [inline v22.1/cockroach-sql-binary.md:10]: Just keep in mind to remove this line when you clone to v22.2.

File: v22.1/cockroach-sql-binary.md
Comment (nickvigilante) [inline v22.1/cockroach-sql-binary.md:18]: Same nit

File: v22.1/cockroach-sql-binary.md
Comment (nickvigilante) [inline v22.1/cockroach-sql-binary.md:147]: Same nit

File: v22.1/cockroach-sql-binary.md
Comment (nickvigilante) [inline v22.1/cockroach-sql-binary.md:168]:
```suggestion
# Server version: CockroachDB CCL {{page.release_info.version}} (x86_64-apple-darwin17.7.0, built {{page.release_info.build_time}}) (same version as client)
```

**Review by nickvigilante — CHANGES_REQUESTED** [review]
That was a long one. See my comments and reach out if you have any questions.

**Review by knz — COMMENTED** [review]
> Reviewed 3 of 5 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/14671)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @gemma-shay, @ianjevans, and @nickvigilante)

___
*[`_includes/v22.1/sql/sql-examples.md` line 404 at r3](https://reviewable.io/reviews/cockroachdb/docs/14671#-N9v39Sp4CjC8qTAMWOi:-N9v39SqEEQ_dPpDJycE:b-r6o4ur) ([raw file](https://github.com/cockroachdb/docs/blob/d42e2575e9e1e58b9ca48aaa9125e57f9e79a0c8/_includes/v22.1/sql/sql-examples.md#L404)):*
> ```Markdown
> ~~~
> 
> ### Edit SQL statements in an external editor
> ```

Please do not include this section in the doc, as we're going to remove this functionality imminently. (https://github.com/cockroachdb/cockroach/pull/86457)

However, separately from this doc project, I'd like to hear your opinion on this: is the ability to use an external editor desirable? Do you think users will be grateful if we streamline this a little bit?
___
*[`_includes/v22.1/sql/sql-examples.md` line 595 at r3](https://reviewable.io/reviews/cockroachdb/docs/14671#-N9v3a7wAA_25a6PiMbL:-N9v3a7wAA_25a6PiMbM:bx1w2im) ([raw file](https://github.com/cockroachdb/docs/blob/d42e2575e9e1e58b9ca48aaa9125e57f9e79a0c8/_includes/v22.1/sql/sql-examples.md#L595)):*
> ```Markdown
> ~~~ shell
> $ cockroach-sql --url='postgres://root@?host=/tmp&port=26257'
> ~~~
> ```

missing newline at end of file
___
*[`v22.1/cockroach-sql.md` line 17 at r3](https://reviewable.io/reviews/cockroachdb/docs/14671#-N9v3ctEAFb3137SCCe5:-N9v3ctEAFb3137SCCe6:b-tqpqg5) ([raw file](https://github.com/cockroachdb/docs/blob/d42e2575e9e1e58b9ca48aaa9125e57f9e79a0c8/v22.1/cockroach-sql.md#L17)):*
> ```Markdown
> {{site.data.alerts.end}}
> 
> The output of `cockroach sql` is part of a stable interface, and can be used programmatically, with the exception of informational output lines that begin with the hash symbol (`#`). Informational output can change from release to release, and should not be used programmatically.
> ```

The output of `cockroach sql` *when used non-interactively* is part of ...
___
*[`v22.1/cockroach-sql-binary.md` line 20 at r3](https://reviewable.io/reviews/cockroachdb/docs/14671#-N9v3lky522qppB_tqSG:-N9v3lky522qppB_tqSH:b-wrwbwc) ([raw file](https://github.com/cockroachdb/docs/blob/d42e2575e9e1e58b9ca48aaa9125e57f9e79a0c8/v22.1/cockroach-sql-binary.md#L20)):*
> ```Markdown
> To exit the interactive shell, use `\q`, `quit`, `exit`, or `ctrl-d`.
> 
> The output of `cockroach-sql` is part of a stable interface, and can be used programmatically, with the exception of informational output lines that begin with the hash symbol (`#`). Informational output can change from release to release, and should not be used programmatically.
> ```

The output of `cockroach sql` *when used non-interactively* is part of ...

(the structure of the input and output during interactive use is unstable)

File: _includes/v22.1/sql/sql-examples.md
Comment (ianjevans) [reply _includes/v22.1/sql/sql-examples.md]: Resolved to **Ctrl+C** in the interim.

File: v22.1/cockroach-sql-binary.md
Comment (ianjevans) [reply v22.1/cockroach-sql-binary.md]: Done.

File: v22.1/cockroach-sql-binary.md
Comment (ianjevans) [reply v22.1/cockroach-sql-binary.md:147]: Done.

**Review by knz — APPROVED** [review]
:lgtm:

> Reviewed 2 of 3 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/14671#-:-NA5esPSCQIVIDCNaaOJ:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @gemma-shay, @ianjevans, and @nickvigilante)

File: _includes/v22.1/sql/sql-examples.md
Comment (ianjevans) [reply _includes/v22.1/sql/sql-examples.md:551]: > > ```gfm
> > ~~~
> > 
> > ### Edit SQL statements in an external editor
> > ```
> 
> Please do not include this section in the doc, as we're going to remove this functionality imminently. ([cockroachdb/cockroach#86457](https://github.com/cockroachdb/cockroach/pull/86457))

Removed.

> However, separately from this doc project, I'd like to hear your opinion on this: is the ability to use an external editor desirable? Do you think users will be grateful if we streamline this a little bit?

I think the usability of the internal editor should be addressed first. If users are breaking out of the shell to edit their SQL externally there are likely two reasons: 1) they find it too cumbersome to use the shell or 2) they are editor absolutists. We can't do much about 2), but we can address the usability of the shell for group 1). Syntax highlighting, and code completion would go a long way to making it easier to accomplish their tasks without needing an external editor.

> missing newline at end of file

Done.

> The output of `cockroach sql` _when used non-interactively_ is part of ...

Done.

**Review by nickvigilante — APPROVED** [review]

**Pull request merged**
